federatedrec/factorization_machine/homo_factorization_machine/homo_fm_guest.py

```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+#
+#  Copyright 2019 The FATE Authors. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import functools
+import numpy as np
+
+from arch.api.utils import log_utils
+from federatedml.framework.homo.procedure import aggregator
+from federatedrec.factorization_machine.fm_model_weight import FactorizationMachineWeights
+from federatedrec.factorization_machine.homo_factorization_machine.homo_fm_base import HomoFMBase
+from federatedml.model_selection import MiniBatch
+from federatedrec.optim.gradient.homo_fm_gradient import FactorizationGradient
+from federatedml.util import consts
+from federatedml.util import fate_operator
+
+LOGGER = log_utils.getLogger()
+
+
+class HomoFMGuest(HomoFMBase):
+    def __init__(self):
+        super(HomoFMGuest, self).__init__()
+        self.gradient_operator = FactorizationGradient()
+        self.loss_history = []
+        self.role = consts.GUEST
+        self.aggregator = aggregator.Guest()
+
+    def _init_model(self, params):
+        super()._init_model(params)
+
+    def fit(self, data_instances, validate_data=None):
+
+        self._abnormal_detection(data_instances)
+        self.init_schema(data_instances)
+
+        validation_strategy = self.init_validation_strategy(data_instances, validate_data)
+        self.model_weights = self._init_model_variables(data_instances)
+
+        mini_batch_obj = MiniBatch(data_inst=data_instances, batch_size=self.batch_size)
+        model_weights = self.model_weights
+
+        degree = 0
+        while self.n_iter_ < self.max_iter:
+            LOGGER.info("iter:{}".format(self.n_iter_))
+            batch_data_generator = mini_batch_obj.mini_batch_data_generator()
+
+            self.optimizer.set_iters(self.n_iter_)
+            if self.n_iter_ > 0 and self.n_iter_ % self.aggregate_iters == 0:
```

Comment:
This one make an aggregation if True. However, after updated last iteration, the updated models are not aggregated. Suggest to add a final aggregation in the last iteration.